students/csdotson/lesson04/trigrams.py

```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/env python3
+# Lesson 4 - Trigrams
+import random
+
+def run():
+    input_file = input("Please enter the name of a file (with extension):\n")
+    read_file(input_file)
+
+
+def read_file(input_file):
+    # Read input_file, scrub for punctuation, create array of words/punctuation
+    with open(input_file, 'r') as f:
+        text = f.read()
+        for entry in punctuation_remove:
+            text = text.replace(entry, '')
+        for entry in replace_with_space:
+            text = text.replace(entry, ' ')
+        for entry in other_punctuation:
+            text = text.replace(entry, f" {entry}")
+        for entry in ending_punctuation:
+            text = text.replace(entry, f" {entry}")
+        create_hash(text.split())
+
+
+def create_hash(text):
+    # Create hash needed to produce trigrams
+    text_hash = {}
+    for index in range(len(text) - 1):
+        key = ' '.join(text[index:index+2])
+        text_hash[key] = []
+    for index in range(len(text)-2):
+        pattern = text[index:index+2]
+        text_hash[' '.join(pattern)].append(text[index+2])
```

Comment:
I'm not sure why you need the two for loops here, should be able to create the look up with just one.

does this have something to do with `if text_hash[next_key] == []`?

if you're trying to safely to exit then you can `if not text_hash.get(next_key) or`